### ISSUE TITLE
Allow extension to configure diagnostic provider: auto, linter, syntax-check or disable

### ISSUE THREAD

**Issue by cedsana** [ansible/vscode-ansible#528]
### Problem

I have the following setup:
- `ansible`
    - `roles`
    - `dev`
        - `roles`
        - `dev.yml`
    - `ansible.cfg`

Whenever I execute the `dev/dev.yml` playbook, I'm always located inside the `ansible` folder.  
Inside my `ansible.cfg`, I have extended the `roles_path` to tell ansible to look for roles inside the `./roles` as well as the `./dev/roles`.  
My `dev/dev.yml` playbook uses roles from both folders and it works perfectly fine.  

However, when I try to edit the `dev/dev.yml` file in VSCode, I get an error stating that the roles cannot be found:

```
Command failed: ansible-playbook /myrepo/ansible/dev/dev.yml --syntax-check
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that
the implicit localhost does not match 'all'
ERROR! the role 'myrole' was not found in <default paths>
```

### Solution

From what I found by searching this repository, the README and the settings of the extension, there is no way to configure the behavior of syntax checking when it's done using `ansible-playbook`.
I would like to either disable the checking entirely or be able to change the configuration to tell it to take my `ansible.cfg` into account. Or simply extend the `role_paths` directly even if it's not DRY.

### Alternatives

Disabling the Ansible extension entirely.  
Attempted to switch to `ansible-lint` but ticking the checkbox didn't change anything.  
Globally specifying the variable `ANSIBLE_CONFIG` but it affects all projects.  
Move the `ansible.cfg` to the root of the VSCode workspace but it will conflict with other teams.

### Additional context

The issue https://github.com/ansible/vscode-ansible/issues/520 was closed.

**Comment by ssbarnea** [comment]
I am afraid that a syntax check failure is almost guaranteed a sign of poof playbook coding, as it does not check for undefined variables. Never assume that user always call a playbook with some variables, check for their presence inside the playbook and fail fast if missing, and you will also pass syntax check.

Regarding the second aspect, supporting `ansible.cfg` files in subdirectories instead of workspace root, that is a very tricky one. We could probably add an option to mention from which subdirectory to run but that would always be limited to one ansible.cfg file per workspace.

If I understood correctly, you have multiple `ansible.cfg` files hosted under a single repository. That sounds like a weird setup and I am not sure if we can make it work with more than one config.

For ansible-lint it might try to call it multiple times for each sub-projects but from the extension point of view, I do not see how it will be addressed. I never seen an IDE extension that allows something like this. Imagine that someone would ask to have a different python virtualenv for each subdirectory inside a repository. I am confident that such a feature request would quickly be closed as wontfix because the complexities involved make it unreasonable to implement or maintain.

@cidrblock @ganeshrn WDYT about that?

**Comment by cedsana** [comment]
Another solution that would work would be to be able to completely disable syntax checking instead.

**Comment by ssbarnea** [comment]
This is not possible, even if you manage to patch the linter to make it skip this step you will discover that everything after falls apart. If syntax cannot pass, you cannot really lint. The only workaround is to skip entire file if you cannot make it work, or do it temporary until you have time to refactor your code to make it pass.

**Comment by cedsana** [comment]
Yes, my bad, I was thinking about disabling linting completely since that's what prints red in all our playbooks.  
Disabling the setting `ansible lint` has no effect since I still get the syntax error.

**Comment by ganeshrn** [comment]
@cedsana
> Another solution that would work would be to be able to completely disable syntax checking instead.

This does defeat the purpose of extension as diagnostics is on of the basic feature that one would expect from extension.  

@bendem
> linting a/playbooks/setup.yml would have a cwd of a, linting b/c/playbooks/setup.yml would have cwd of b/c, linting anything in c would have a cwd of ..

> If the linting is based on a server, this is harder since servers generally run from a single directory. You'd run multiple servers I guess.

When a new workspace is added in extension the [ansible-language-server](https://github.com/ansible/ansible-language-server) does a `anisble-config dump` from workspace root folder and while providing diagnostics the `ansible-lint` and `ansible-playbook --syntax-check` commands with workspace root as CWD. We might have to check if `ansible.cfg` file is present in the working directory of the opened document and update the AnsibleConfig service everytime the file is edited and saved which might result in a delay in providing the diagnostics information.

**Comment by cidrblock** [comment]
The only 2 things I can think of here:
1) Open the parent ansible directory as the workspace root
2) Set the environment variable for the vscode session with an .env file: 
```
The extension also loads an [environment variable definitions file](https://code.visualstudio.com/docs/python/environments#_environment-variable-definitions-file) identified by the python.envFile setting. The default value of this setting is ${workspaceFolder}/.env.
```

https://code.visualstudio.com/docs/python/environments

I wonder if either of these will work for you?

**Comment by cedsana** [comment]
@ganeshrn 
> This does defeat the purpose of extension as diagnostics is on of the basic feature that one would expect from extension.

I would argue that the extension is mainly downloaded and installed for its syntax highlighting capabilities since it's the only existing thing capable of solving this for users.  
That's also the reason why I keep using it even though I get flooded with "syntax errors" everywhere in my playbooks.  

Even if you were right, I feel like it would only enforce the suggestion of this issue: if the main feature of this extension is to provide syntax formatting, it seems natural that adding a setting for making it more flexible is the way forward.



@cidrblock 
> Set the environment variable for the vscode session with an .env file

What you are suggesting is interesting because that's the reason I created this issue. At first I planned on finding way to specify an environment variable for an extension and stumbled on this issue https://github.com/microsoft/vscode/issues/91372 that said "This should be solved with a setting in the extension itself".

**Comment by ssbarnea** [comment]
@cedsana Let me explain you why there is no setting inside the app itself. Assume that there is one and it works the way you want. After this you try to run `ansible-lint` from vscode terminal, or another terminal and you get surprised that you get different outcomes than inside vscode problems tab.

By ensuring that ansible itself can find the config (ANSIBLE_CONFIG), you address all use-cases, in terminal, in vscode terminal and language-server. A configuration option in vscode, will not be able to fix ansible(-lint) in your terminal.

AFAIK, our language-server is able to even detect which ansible.cfg is used and it does that by running `ansible-config dump`. Alternatively you can also run `ansible --version` as that does also include this information.

Basically we want to avoid adding an option that would only create more confusion for the users.

**Comment by ganeshrn** [comment]
> One might argue our repository structure is weird but it simply is a monorepo in which I work on several devops-related projects at once.

The extension supports [multiroot workspaces](https://code.visualstudio.com/docs/editor/multi-root-workspaces). So I think you can setup the VSCode window by adding mulitple folders (projects) in the window `File -> Add folder to workspace` 


>. Enable ansible-lint setting. I expected it to disable ansible-playbook --syntax-check as well.

Diagnostic through ansible-lint is optional as it explicitly required installing `ansible-lint`. If it is not installed the fallback is to `ansible-playbook --syntax-check` 


>. That's also the reason why I keep using it even though I get flooded with "syntax errors" everywhere in my playbooks.

This can be avoided by using multi-root workspaces.

**Comment by ganeshrn** [comment]
Fixed by PR https://github.com/ansible/vscode-ansible/pull/648 and https://github.com/ansible/ansible-language-server/pull/448

Please update the extension to the latest pre-release version and try it out.
